### PR TITLE
Clean block queue code

### PR DESCRIPTION
**Pull request by pdobacz**
## Overview

BlockQueue.Core (especially tests) aged a bit, which I spotted when solving #1136. This should give those a significant refactor.

As far as I can tell, this shouldn't change `BlockQueue`'s behavior in any way.

## Changes

This is a set of interleaving implementation and test changes. It might be **most convenient** to review commit-by-commit or in small batches
- **removing some dead code**
  - refactor: remove BlockQueue.Core.new/0 which wasn't really used
  - refactor: get rid of child_start_parent_height handling
- **mechanical and simple tidies to `CoreTests`**
  - test: remove duplicate test and tighten the one that remained after
  - style: rearrange existing BlockQueue.CoreTests into function `describe` blocks
  - refactor: ditch the recover/3 helper in test
- **larger refactor cleaning `Core.new/1` that had a multitude of named args**
  - refactor: hideoptional config parameters in tests and make consistent …
- **tidies of tests now grouped in the new `describe` blocks, adding some new ones** 
  - test: tidy new/1 tests
  - style: replace the confusing x * @child_block_interval DRYing
  - test: tidy enqueue_block and set_ethereum_height tests
  - test: simple tidies of get_blocks_to_submit/1 tests
- **large refactor/revamp of gas price tests** (actually, this was my biggest itch in here. Those tests used a lot of internal stuffs to test uncleanly. Now they are tested based on public API calls and observable actions like ethereum progressing/new blocks etc. 
  - refactor: change gas price tests to not use internals

**NOTE** on the gas price test refactors. This actually uncovered a bunch of surprising behaviors, so careful review required on the last commit :point_up: . So far I haven't attempted to "fix" this, since there hasn't been an issue with the mechanism as a whole. For now I only solidified existing behavior in the tests.

Overall, this aims to get closer to the `ExUnit` style we have, reduce the use of test helpers and make tests more approachable (hopefully).

## Testing

`mix test test/omg_child_chain` and :eyes:

### CODE REVIEW

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs [hunk 1]
```diff
@@ -29,50 +30,27 @@ defmodule OMG.ChildChain.BlockQueue.CoreTest do
...
         child_block_interval: @child_block_interval,
-        chain_start_parent_height: 1,
-        block_submit_every_nth: 1,
-        finality_threshold: 12,
-        last_enqueued_block_at_height: 0
+        gas_price_adj_params: %GasPriceAdjustment{max_gas_price: 40_000_000_000}
       )
 
-    empty_with_gas_params = %{empty | formed_child_block_num: 5 * @child_block_interval, gas_price_to_use: 100}
+    empty_with_gas_params = %{empty | formed_child_block_num: 5000, gas_price_to_use: 100}
```
Comment (pnowosie) [inline apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: Why? IMHO the former was more descriptive.

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs [hunk 2]
```diff
@@ -29,50 +30,27 @@ defmodule OMG.ChildChain.BlockQueue.CoreTest do
...
 
-    {:do_form_block, empty_with_gas_params} =
-      Core.set_ethereum_status(empty_with_gas_params, 1, 3 * @child_block_interval, false)
+    {:do_form_block, empty_with_gas_params} = Core.set_ethereum_status(empty_with_gas_params, 1, 3000, false)
 
     # assertions - to be explicit how state looks like
-    child_block_mined = 3 * @child_block_interval
+    child_block_mined = 3000
```
Comment (pnowosie) [inline apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: Why? IMHO the former was more descriptive.
(and many more :point_down: in the file)

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs (hunk 1 above)
Comment (achiurizo) [reply apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: I second this. It's much clearer to understand what this value represents with the `@child_block_interval`.

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs [hunk 3]
```diff
@@ -29,50 +30,27 @@ defmodule OMG.ChildChain.BlockQueue.CoreTest do
...
-      block_submit_every_nth: 1,
-      finality_threshold: 12,
-      last_enqueued_block_at_height: 0
-    )
+    {:ok,
+     %{empty: empty, empty_with_gas_params: empty_with_gas_params, empty_high_max_gas_price: empty_high_max_gas_price}}
   end
 
-  describe "Block queue." do
+  describe "child_block_nums_to_init_with/4" do
```
Comment (achiurizo) [inline apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: ✨

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs [hunk 4]
```diff
@@ -89,173 +67,124 @@ defmodule OMG.ChildChain.BlockQueue.CoreTest do
...
-          parent_height: 6,
-          child_block_interval: @child_block_interval,
-          chain_start_parent_height: 1,
-          block_submit_every_nth: 1,
-          finality_threshold: finality_threshold,
-          last_enqueued_block_at_height: 0
+          6000,
+          [{5000, "5"}, {6000, "6"}, {7000, "7"}, {8000, "8"}, {9000, "9"}],
+          "6",
+          10,
```
Comment (achiurizo) [inline apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: Is there a reason we moved away from the named arguments? I think it helped make it clear what the arguments represented and how to invoke `Core.new` here.

File: apps/omg_child_chain/lib/omg_child_chain/block_queue.ex [hunk 5]
```diff
@@ -80,15 +83,13 @@ defmodule OMG.ChildChain.BlockQueue do
...
-                 known_hashes: Enum.zip(range, known_hashes),
-                 top_mined_hash: top_mined_hash,
-                 parent_height: parent_height,
+                 mined_num,
+                 Enum.zip(range, known_hashes),
+                 top_mined_hash,
+                 parent_height,
                  child_block_interval: child_block_interval,
-                 chain_start_parent_height: parent_start,
                  block_submit_every_nth: Application.fetch_env!(:omg_child_chain, :block_submit_every_nth),
```
Comment (InoMurko) [inline apps/omg_child_chain/lib/omg_child_chain/block_queue.ex]: A good pattern coming from OTP is that children never directly access configuration. Configuration values are always passed top to bottom. Application process to Supervisor to it's children.

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs (hunk 3 above)
Comment (InoMurko) [reply apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: now we need an explanation what the function name means :D

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs (hunk 4 above)
Comment (pdobacz) [reply apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: ah, so I liked those named args here too, but:
 - there were overall too many of those, and I wanted a clear and standard distinction between required and optional
 - I tried sth like `def new([required_1: ..., required_2: ... | optional])` to have best of both, but it seems not not be supported and the alternative: `def new([{:required_1, _}, {:required_2, _} | optional` either didn't work or it seemed weird and nonstandard. 
 - And then just `def new(opts \\ [])` and have the required args enforced in the body didn't click too well neither. But maybe this could be an alternative, it just would hide the required vs optional distinction.

I think 4 unnamed required params and then options is pretty standard. I can try the last `def new(opts \\ [])` if you like and see how it goes.

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs [hunk 6]
```diff
@@ -19,6 +19,7 @@ defmodule OMG.ChildChain.BlockQueue.CoreTest do
   import ExUnit.CaptureLog
```
Comment (InoMurko) [inline apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: I think a pattern that is quite good and I usually follow is to name the argument variables I'm using:
```
 Core.child_block_nums_to_init_with(0, 3000, @child_block_interval, 2)
```
I would call this:
```
mined_num = 0
until_child_block_num = 300, 
finality_threshold = 2
 Core.child_block_nums_to_init_with(mined_num, until_child_block_num, @child_block_interval, finality_threshold)
```
using this pattern, I do not need to store in memory what these arguments are or what they mean. I think we should follow this...

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs [hunk 7]
```diff
@@ -340,164 +348,218 @@ defmodule OMG.ChildChain.BlockQueue.CoreTest do
...
-      {:do_form_block, newstate} =
-        state
-        |> Core.enqueue_block(<<0>>, 6000, 1)
-        |> Core.set_ethereum_status(1 + eth_gap, 0, false)
-
-      assert current_price < newstate.gas_price_to_use
-
-      # assert the actual gas price based on parameters value - test could fail if params or calculation will change
-      assert 200 == newstate.gas_price_to_use
+    test "Progressing to enqueue height may raise price if threshold exceeded", %{empty_high_max_gas_price: empty} do
```
Comment (pnowosie) [inline apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: I had a hard time to understand test description, especially `may` & `if`
I'd like sth with more assurance ~ `exceeding the threshold height results with price raise`

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs [hunk 8]
```diff
@@ -340,164 +348,218 @@ defmodule OMG.ChildChain.BlockQueue.CoreTest do
...
-      {_, state} = Core.set_ethereum_status(state, 101, 0, false)
-      assert state.gas_price_to_use > gas_price
-    end
-
-    test "gas price changes only, when etherum advanses", %{empty_with_gas_params: state} do
-      gas_price = state.gas_price_to_use
-      eth_height = 0
+      assert [%{gas_price: 5}, %{gas_price: 5}] = Core.get_blocks_to_submit(state)
 
+      # no more successful submissions can change the price
```
Comment (pnowosie) [inline apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: I don't follow the test logic atm. If I'm correct you are doing the exact same thing 2 more times (after prev `1000-2` trials) with a comment that now it can change the price. Final asserts shows the price hasn't changed :question:

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs (hunk 7 above)
Comment (pdobacz) [reply apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: um, so the test and description are as un-assuring as the behavior. It means that despite the `set_ethereum_status` goes to exactly `enqueue_block`'s eth_height, it still raises the price.

I'll clarify to

```
    # NOTE: This behavior is strange - revisit along with the other ones (see `NOTEs` here)
    #       Why would detecting the exact height of enqueue cause the price to raise - we don't know if there's need to
    test "Despite progressing to enqueue height price may raise", %{empty_high_max_gas_price: empty} do
```

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs (hunk 8 above)
Comment (pdobacz) [reply apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: yeah, so the idea is to "reach" the bottom price (5) and do one more (actually I don't know why I did 2, a leftover, I'll fix this) and see it remain at 5

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs (hunk 4 above)
Comment (pdobacz) [reply apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: ok, did that, I think it looks better ff2a127946a4c8de40a752a1d2827674056c4788, thx!

File: apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs (hunk 1 above)
Comment (pdobacz) [reply apps/omg_child_chain/test/omg_child_chain/block_queue/core_test.exs]: I tried this too, but it really is unreadable, with cases like: 

```
      assert [2 * @child_block_interval, 3 * @child_block_interval] ==
               Core.child_block_nums_to_init_with(
                 2 * @child_block_interval,
                 3 * @child_block_interval,
                 @child_block_interval,
                 0
               )
```

and sometimes one can't pattern-match on the `n * ...` expression in the result. I opt towards leaving it consistently like:
 - when interval specifically is meant (as in a config param passed in) - `@child_block_interval`
 - when a blknum is meant `<n>000`.

(I have a full commit doing the change in a branch in case you're interested).

WDYT?

**Comment by pdobacz** [comment]
OK, since there's :heavy_check_mark: 's and no further folluw-up comments in the style-related threads, merging this.

I'm going to open an issue to dig deeper into the strange behaviors spotted in the tests (everything seemed to work but maybe some edge cases could be better handled, when doing gas price selection).
